src/porepy/numerics/ad/forward_mode.py

```diff
@@ -91,24 +91,84 @@ def __rmul__(self, other):
 
     def __pow__(self, other):
         if not isinstance(other, Ad_array):
-            val = self.val**other
-            jac = self.diagvec_mul_jac(other * self.val ** (other - 1))
+            if isinstance(other, int) or isinstance(other, np.integer):
+                # Standard ints and numpy scalars of integer format can be converted to
+                # float in a standard way
+                val = self.val ** float(other)
+                jac = self.diagvec_mul_jac(other * self.val ** float(other - 1))
+            elif isinstance(other, np.ndarray) and np.issubdtype(
+                other.dtype, np.integer
+            ):
+                # Numpy arrays of integer format are converted using np.astype
+                val = self.val ** other.astype(float)
+                jac = self.diagvec_mul_jac(
+                    other * self.val ** (other.astype(float) - 1)
+                )
+            else:
+                # Other should be a float, or have float data type; raising to the power
+                # of other should anyhow be fine. If there are more special cases not
+                # yet hit upon, an error message will be given here.
+                val = self.val**other
+                jac = self.diagvec_mul_jac(other * self.val ** (other - 1))
+
```

Comment:
Couldn't case 1 and 3 be handled as one (with the solution for 1, which doesn't do anything for 3)? Don't know if it's better

Reply:
Unfortunately no . `other` can be a np.array, in which case `float(other)` gives an error (it turned out). One could let the second condition find all `numpy `arrays (and do an unnecessary conversion if `dtype=float`), but the third else would still be needed for `floats`, and other data types for which the operation works out of the box. It also turned out, there are more special cases in the form of data types than what I would have thought.